File: baselib/src/main/java/com/stjy/baselib/base/BaseFragment.java
package com.stjy.baselib.base;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.alibaba.android.arouter.launcher.ARouter;
import com.blankj.utilcode.util.ActivityUtils;
import com.blankj.utilcode.util.PermissionUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.qmuiteam.qmui.widget.dialog.QMUIDialog;
import com.qmuiteam.qmui.widget.dialog.QMUITipDialog;
import com.stjy.baselib.R;
import com.stjy.baselib.listener.PermissionListener;
import com.stjy.baselib.utils.EventBusUtils;
import com.stjy.baselib.utils.RxLifecycleUtils;
import com.stjy.baselib.wigiet.stateview.StateView;
import com.tbruyelle.rxpermissions2.RxPermissions;
import com.trello.rxlifecycle2.LifecycleTransformer;
import com.zhouyou.http.subsciber.IProgressDialog;

import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import me.yokeyword.fragmentation.SupportFragment;

import static com.qmuiteam.qmui.widget.dialog.QMUITipDialog.Builder.ICON_TYPE_LOADING;

/**
 * @author DaiFalin
 */
public abstract class BaseFragment extends SupportFragment implements View.OnClickListener {

    protected Context mContext;
    protected View mView;
    protected StateView mStateView;
    protected CompositeDisposable mDisposablePool = new CompositeDisposable();
    private BaseActivity mActivity;
    private TextView mBarTitle;
    private TextView mBarRight;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            this.mContext = context;
            this.mActivity = (BaseActivity) context;
        }
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            this.mContext = activity;
            this.mActivity = (BaseActivity) activity;
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ARouter.getInstance().inject(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        mView = inflater.inflate(getLayoutID(), container, false);
        mStateView = StateView.inject(mView);
        return mView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initToolBar();
        initView(mView);
        initData();
        initListener();
    }

    @Override
    public void onStart() {
        super.onStart();
        if (isRegisterEvent()) {
            EventBusUtils.register(this);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (isRegisterEvent()) {
            EventBusUtils.unregister(this);
        }
        if (mActivity.mProgressDialog != null) {
            mActivity.mProgressDialog.dismiss();
        }
        mDisposablePool.clear();
    }

    /**
     * 是否订阅Bus事件
     *
     * @return
     */
    public boolean isRegisterEvent() {
        return false;
    }

    /**
     * 初始化ToolBar
     */
    private void initToolBar() {
        mBarTitle = mView.findViewById(R.id.bar_title);
        mBarRight = mView.findViewById(R.id.bar_right);
        //判断是否有Toolbar,并默认显示返回按钮
        if (null != getToolbar() && isShowBacking()) {
            setNavigationIcon(R.mipmap.go_back_black);
            getToolbar().setNavigationOnClickListener(v -> {
                setNavigationOnClickListener();
            });
        }
    }

    public void setNavigationOnClickListener() {
        mActivity.finish();
    }

    /**
     * 设置头部标题
     *
     * @param title
     */
    public void setBarTitle(CharSequence title) {
        if (null != mBarTitle) {
            mBarTitle.setText(title);
        }
    }

    /**
     * 设置右边的文字
     *
     * @return
     */
    public void setBarRightText(CharSequence subTitle) {
        if (null != mBarRight) {
            mBarRight.setText(subTitle);
        }
    }

    /**
     * 设置右边的文字的点击监听
     *
     * @param listener
     */
    public void setOnClickRightTextListener(View.OnClickListener listener) {
        if (null != mBarRight && null != listener) {
            mBarRight.setOnClickListener(listener);
        }
    }

    /**
     * this Activity of tool bar.
     * 获取头部.
     *
     * @return support.v7.widget.Toolbar.
     */
    public Toolbar getToolbar() {
        return mView.findViewById(R.id.toolbar);
    }

    protected <T> LifecycleTransformer<T> bindLifecycle() {
        return RxLifecycleUtils.bindLifecycle(this);
    }

    /**
     * 后退按钮图片
     */
    public void setNavigationIcon(@DrawableRes int resId) {
        getToolbar().setNavigationIcon(resId);
    }

    /**
     * 是否显示后退按钮,默认显示,可在子类重写该方法.
     *
     * @return
     */
    public boolean isShowBacking() {
        return true;
    }

    /**
     * 打开新的Activity
     *
     * @param toActivity 新Activity
     */
    public void startActivity(Class<? extends AppCompatActivity> toActivity) {
        startActivity(toActivity, false);
    }

    /**
     * 打开新的Activity
     *
     * @param toActivity 新Activity
     * @param isFinish   是否关闭当前Activity
     */
    public void startActivity(Class<? extends AppCompatActivity> toActivity, boolean isFinish) {
        ActivityUtils.startActivity(getBaseActivity(), toActivity);
        if (isFinish) {
            getBaseActivity().finish();
        }
    }

    @Override
    public void startActivityForResult(Intent intent, int requestCode) {
        Fragment fragment = getParentFragment();
        if (fragment != null) {
            fragment.startActivityForResult(intent, requestCode);
        } else {
            super.startActivityForResult(intent, requestCode);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        List<Fragment> fragmentList = getChildFragmentManager().getFragments();
        if (fragmentList != null) {
            for (Fragment fragment : fragmentList) {
                fragment.onActivityResult(requestCode, resultCode, data);
            }
        }
    }

    public void startLoadingDialog() {
        mActivity.startLoadingDialog();
    }

    public void stopLoadingDialog() {
        mActivity.stopLoadingDialog();
    }

    public void startLoading() {
        if (mStateView != null) {
            mStateView.showLoading();
        }

    }

    public void stopLoading() {
        if (mStateView != null) {
            mStateView.showContent();
        }
    }

    public IProgressDialog getIProgressDialog() {
        return this::getProgressDialog;
    }

    public Dialog getProgressDialog() {
        return new QMUITipDialog.Builder(getBaseActivity())
                .setIconType(ICON_TYPE_LOADING)
                .setTipWord("加载中")
                .create(false);
        /*return new MaterialDialog.Builder(getBaseActivity())
                .content("加载中...")
                .cancelable(false)
                .canceledOnTouchOutside(false)
                .progress(true, 0)
                .build();*/
    }

    @Override
    public void onClick(View v) {

    }

    /**
     * 获取BaseActivity对象
     *
     * @return BaseActivity对象
     */
    public BaseActivity getBaseActivity() {
        return mActivity;
    }

    public View getFakeStatusBar() {
        return mView.findViewById(R.id.fake_status_bar);
    }

    /**
     * 请求权限
     *
     * @param permissions
     */
    @SuppressLint("CheckResult")
    public void requestPermission(PermissionListener listener, @NonNull String... permissions) {
        RxPermissions rxPermissions = new RxPermissions(this);
        rxPermissions.requestEachCombined(permissions)
                .subscribe(permission -> {
                    if (permission.granted) {
                        // All permissions are granted !
                        listener.onGranted();
                    } else if (permission.shouldShowRequestPermissionRationale) {
                        // At least one denied permission without ask never again
                        ToastUtils.showShort("权限被拒绝");
                    } else {
                        // At least one denied permission with ask never again
                        // Need to go to the settings
                        new QMUIDialog.MessageDialogBuilder(getBaseActivity())
                                .setMessage("权限被拒绝并设置为不再询问，请前往设置中开启")
                                .addAction("去设置", (dialog, index) -> PermissionUtils.launchAppDetailsSettings())
                                .addAction("下次再说", (dialog, index) -> dialog.dismiss())
                                .create()
                                .show();
                    }
                });
    }

    /**
     * 布局的LayoutID
     *
     * @return LayoutID
     */
    protected abstract int getLayoutID();

    /**
     * 初始化子View
     *
     * @param contentView view
     */
    protected abstract void initView(View contentView);

    /**
     * 初始化数据
     */
    protected abstract void initData();

    /**
     * 初始化监听器
     */
    protected abstract void initListener();

}

File: baselib/src/main/java/com/stjy/baselib/base/mvc/BaseFragment.kt
package com.stjy.baselib.base.mvc

import android.annotation.SuppressLint
import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.CheckResult
import androidx.annotation.DrawableRes
import androidx.annotation.NonNull
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.alibaba.android.arouter.launcher.ARouter
import com.blankj.utilcode.util.ActivityUtils
import com.gyf.immersionbar.ImmersionBar
import com.gyf.immersionbar.components.SimpleImmersionOwner
import com.gyf.immersionbar.components.SimpleImmersionProxy
import com.gyf.immersionbar.ktx.immersionBar
import com.stjy.baselib.R
import com.stjy.baselib.utils.EventBusUtils
import com.stjy.baselib.utils.RxLifecycleUtils
import com.stjy.baselib.wigiet.stateview.StateView
import com.trello.rxlifecycle2.LifecycleProvider
import com.trello.rxlifecycle2.LifecycleTransformer
import com.trello.rxlifecycle2.RxLifecycle
import com.trello.rxlifecycle2.android.FragmentEvent
import com.trello.rxlifecycle2.android.RxLifecycleAndroid
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import me.yokeyword.fragmentation.SupportFragment


/**
 * @Author: superman
 * @CreateTime: 2020/7/4
 * @Describe: BaseFragment基类
 */
abstract class BaseFragment : SupportFragment(), SimpleImmersionOwner, View.OnClickListener , LifecycleProvider<FragmentEvent> {
    private val lifecycleSubject = BehaviorSubject.create<FragmentEvent>()
    private var mSimpleImmersionProxy=SimpleImmersionProxy(this)//ImmersionBar代理类
    lateinit var mActivity: BaseActivity
    private var mView: View? = null
    protected lateinit var mStateView: StateView
    var mDisposablePool = CompositeDisposable()
    private var mBarTitle: TextView? = null
    private var mBarRight: TextView? = null

    override fun onAttach(activity: Activity) {
        lifecycleSubject.onNext(FragmentEvent.ATTACH)
        super.onAttach(activity)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        lifecycleSubject.onNext(FragmentEvent.CREATE)
        super.onCreate(savedInstanceState)
        ARouter.getInstance().inject(this)
        mActivity = activity as BaseActivity
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        lifecycleSubject.onNext(FragmentEvent.CREATE_VIEW)
        mView = with(inflater.inflate(getLayoutID(), container, false)) {
            mStateView = StateView.inject(this)
            this
        }
        return mView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mSimpleImmersionProxy.onActivityCreated(savedInstanceState)
        if (isRegisterEvent()) {
            EventBusUtils.register(this)
        }
        initImmersionBar()//初始化沉浸式
        initToolBar()
        initView(mView)
        initListener()
        initData()
    }

    override fun onStart() {
        lifecycleSubject.onNext(FragmentEvent.START)
        super.onStart()
    }

    override fun onResume() {
        lifecycleSubject.onNext(FragmentEvent.RESUME)
        super.onResume()
    }

    override fun onPause() {
        lifecycleSubject.onNext(FragmentEvent.PAUSE)
        super.onPause()
        stopLoading()
    }

    override fun onStop() {
        lifecycleSubject.onNext(FragmentEvent.STOP)
        super.onStop()
    }

    override fun onDestroyView() {
        lifecycleSubject.onNext(FragmentEvent.DESTROY_VIEW)
        super.onDestroyView()
        mSimpleImmersionProxy.onDestroy()
        if (isRegisterEvent()) {
            EventBusUtils.unregister(this)
        }
        mActivity.mLoadingDialog?.dismissing()
        mDisposablePool.clear()
    }

    override fun onDestroy() {
        lifecycleSubject.onNext(FragmentEvent.DESTROY)
        super.onDestroy()
    }

    override fun onDetach() {
        lifecycleSubject.onNext(FragmentEvent.DETACH)
        super.onDetach()
    }

    @NonNull
    @CheckResult
    override fun lifecycle(): Observable<FragmentEvent> {
        return lifecycleSubject.hide()
    }

    @NonNull
    @CheckResult
    override fun <T : Any> bindUntilEvent(event: FragmentEvent): LifecycleTransformer<T> {
        return RxLifecycle.bindUntilEvent(lifecycleSubject, event)
    }

    @NonNull
    @CheckResult
    override fun <T> bindToLifecycle(): LifecycleTransformer<T> {
        return RxLifecycleAndroid.bindFragment(lifecycleSubject)
    }

    /**
     * 是否订阅Bus事件
     * @return
     */
    open fun isRegisterEvent(): Boolean = false

    /**
     * 初始化ToolBar
     */
    private fun initToolBar() {
        mBarTitle = mView?.findViewById(R.id.bar_title)
        mBarRight = mView?.findViewById(R.id.bar_right)
        //判断是否有Toolbar,并默认显示返回按钮
        toolbar?.let {
            initImmersionBar()
            if (isShowBacking()) {
                setNavigationIcon(R.mipmap.ic_black)
                it.setNavigationOnClickListener { setNavigationOnClickListener() }
            }
        }
    }

    open fun setNavigationOnClickListener() {
        mActivity.finish()
    }

    /**
     * 设置头部标题
     * @param title
     */
    fun setBarTitle(title: CharSequence?) {
        mBarTitle?.text = title
    }

    /**
     * 设置右边的文字
     * @return
     */
    fun setBarRightText(subTitle: CharSequence?) {
        mBarRight?.text = subTitle
    }

    /**
     * 设置右边的文字的点击监听
     * @param listener
     */
    fun setOnClickRightTextListener(listener: View.OnClickListener?) {
        if (null != listener) {
            mBarRight?.setOnClickListener(listener)
        }
    }

    /**
     * this Activity of tool bar.
     * 获取头部
     * @return support.v7.widget.Toolbar.
     */
    val toolbar: Toolbar?
        get() = mView?.findViewById(R.id.toolbar)

    protected fun <T> bindLifecycle(): LifecycleTransformer<T> {
        return RxLifecycleUtils.bindLifecycle(this)
    }

    /**
     * 后退按钮图片
     */
    open fun setNavigationIcon(@DrawableRes resId: Int) {
        toolbar?.setNavigationIcon(resId)
    }

    /**
     * 是否显示后退按钮,默认显示,可在子类重写该方法.
     * @return
     */
    open fun isShowBacking(): Boolean = true

    /**
     * 打开新的Activity
     * @param toActivity 新Activity
     * @param isFinish   是否关闭当前Activity
     */
    @JvmOverloads
    fun startActivity(toActivity: Class<out AppCompatActivity?>, isFinish: Boolean = false) {
        ActivityUtils.startActivity(mActivity, toActivity)
        if (isFinish) {
            mActivity.finish()
        }
    }

    override fun startActivityForResult(intent: Intent?, requestCode: Int) {
        val fragment = parentFragment
        if (fragment != null) {
            fragment.startActivityForResult(intent, requestCode)
        } else {
            super.startActivityForResult(intent, requestCode)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val fragmentList = childFragmentManager.fragments
        if (fragmentList != null) {
            for (fragment in fragmentList) {
                fragment.onActivityResult(requestCode, resultCode, data)
            }
        }
    }

    open fun startLoadingDialog(content: CharSequence? = null,
                                contentTextColor: Int? = null,
                                contentTextSize: Int? = null,
                                cancelable: Boolean? = null,
                                canceledOnTouchOutside: Boolean? = null,
                                strokeWidth: Int? = null,
                                strokeColor: Int? = null,
                                solidColor: Int? = null,
                                cancelListener: DialogInterface.OnCancelListener? = null,
                                cornerRadius: Int? = null,
                                LoadingViewSize: Int? = null,
                                dimAmount: Float? = null) {
        mActivity.startLoadingDialog(content,
                contentTextColor,
                contentTextSize,
                cancelable,
                canceledOnTouchOutside,
                strokeWidth,
                strokeColor,
                solidColor,
                cancelListener,
                cornerRadius,
                LoadingViewSize,
                dimAmount)
    }

    open fun stopLoadingDialog() {
        mActivity.stopLoadingDialog()
    }

    fun startLoading() {
        mStateView.showLoading()
    }

    fun stopLoading() {
        mStateView.showContent()
    }

    override fun onClick(v: View) {}

    /**
     * 布局的LayoutID
     * @return LayoutID
     */
    protected abstract fun getLayoutID(): Int

    /**
     * 初始化子View
     * @param contentView view
     */
    protected abstract fun initView(contentView: View?)

    /**
     * 初始化数据
     */
    protected abstract fun initData()

    /**
     * 初始化监听器
     */
    protected abstract fun initListener()

    /**
     * 扩展点击事件
     */
    fun View.onClick(listener: View.OnClickListener): View {
        setOnClickListener(listener)
        return this
    }

    /**
     *扩展点击事件，参数为方法
     */
    fun View.onClick(method: () -> Unit): View {
        setOnClickListener { method() }
        return this
    }

    /**
     *扩展视图可见性
     */
    fun View.setVisible(visible: Boolean) {
        this.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        mSimpleImmersionProxy.isUserVisibleHint = isVisibleToUser
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        mSimpleImmersionProxy.onHiddenChanged(hidden)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        mSimpleImmersionProxy.onConfigurationChanged(newConfig)
    }

    /**
     * 是否可以实现沉浸式，当为true的时候才可以执行initImmersionBar方法
     * Immersion bar enabled boolean.
     * @return the boolean
     */
    override fun immersionBarEnabled() = true

    /**
     * 初始化沉浸式代码
     * Init immersion bar.
     */
    override fun initImmersionBar() {
        immersionBar {
            statusBarColor(R.color.white)
            navigationBarColor(R.color.white)
            statusBarDarkFont(true)
            navigationBarDarkIcon(true)
            fitsSystemWindows(true)
        }
    }

    /**
     * 请求权限
     * @param permissions
     */
    @SuppressLint("CheckResult")
    fun requestPermission(vararg permissions: String, success: (() -> Unit)? = null, failed: (() -> Unit)? = null) {
        mActivity.requestPermission(*permissions, success = success, failed = failed)
    }

}